# Numpy Project 2:

Weather Data Analysis Using Numpy

In [91]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("rajaisrarkiani/pakistan-historical-weather-data-20002024")

print("Path to dataset files:", path)

Using Colab cache for faster access to the 'pakistan-historical-weather-data-20002024' dataset.
Path to dataset files: /kaggle/input/pakistan-historical-weather-data-20002024


In [92]:
import numpy as np

numeric_data = np.genfromtxt('/kaggle/input/pakistan-historical-weather-data-20002024/pakistan_weather_2000_2024.csv', delimiter=',', dtype=float, skip_header=True)
full_data = np.genfromtxt('/kaggle/input/pakistan-historical-weather-data-20002024/pakistan_weather_2000_2024.csv', delimiter=',', dtype=str)

In [93]:
# headers
header = full_data[0,:]

In [94]:
header

array(['date', 'year', 'month', 'day', 'dayofweek', 'is_weekend',
       'season', 'city', 'region', 'latitude', 'longitude', 'elevation',
       'tmin', 'tmax', 'tavg', 'prcp', 'wspd', 'humidity', 'pressure',
       'dew_point', 'cloud_cover', 'visibility', 'temp_range',
       'is_hot_day', 'is_cold_day', 'rainfall_intensity', 'wind_category'],
      dtype='<U18')

In [95]:
# fetching numerics data in a new array
numeric_data = numeric_data[:,[11,12,13,14,15,16,17,18,19,20,22]]
numeric_data[0,:]

array([ 540. ,    3.1,   17.2,    9.4,    0. ,    6. ,   76. , 1018.9,
          4.8,    0. ,   14.1])

In [96]:
# Checking null values
np.isnan(numeric_data).sum()

np.int64(0)

In [97]:
# Mentioning the numerics columns for further transforming
# latitude,9 longitude,10 elevation,11 tmin,12 tmax,13 tavg,14 prcp,15 wspd,16 humidity,17 pressure,18 dew_point,19 cloud_cover,20 temp_range,22
full_data = full_data[1:,:]

first_string_cols = full_data[:, :9]
first_num_cols = full_data[:, 9:21].astype(float)
sec_num_cols = full_data[:, 22].astype(object)
sec_num_cols = np.where(sec_num_cols == '', np.nan, sec_num_cols)
sec_num_cols = sec_num_cols.astype(float)
sec_string_cols = full_data[:, 23:24]
third_string_cols = full_data[:, 25:]
sec_string_cols = np.where(sec_string_cols == '1', 'Yes', 'No')

full_data = np.hstack((first_string_cols, first_num_cols, sec_num_cols.reshape(-1,1), sec_string_cols, third_string_cols))
full_data

array([['1/1/2000', '2000', '1', ..., 'No', 'none', 'windy'],
       ['1/2/2000', '2000', '1', ..., 'No', 'none', 'windy'],
       ['1/3/2000', '2000', '1', ..., 'No', 'none', 'windy'],
       ...,
       ['12/29/2011', '2011', '12', ..., 'No', 'light', 'breezy'],
       ['12/30/2011', '2011', '12', ..., 'No', 'none', 'breezy'],
       ['12/31/2011', '2011', '12', ..., 'No', 'none', 'breezy']],
      dtype='<U32')

In [98]:
full_data[0:2,:]

array([['1/1/2000', '2000', '1', '1', '5', '1', 'Winter', 'Islamabad',
        'Punjab', '33.6844', '73.0479', '540.0', '3.1', '17.2', '9.4',
        '0.0', '6.0', '76.0', '1018.9', '4.8', '0.0', '14.1', 'No',
        'none', 'windy'],
       ['1/2/2000', '2000', '1', '2', '6', '1', 'Winter', 'Islamabad',
        'Punjab', '33.6844', '73.0479', '540.0', '2.9', '17.4', '9.3',
        '0.0', '6.1', '77.0', '1018.6', '4.8', '0.0',
        '14.499999999999998', 'No', 'none', 'windy']], dtype='<U32')

In [99]:
mean = np.mean(numeric_data, axis=0)
mean

array([ 651.6568174 ,   15.18648793,   25.7875767 ,   20.35544857,
          1.71780106,   12.90230655,   58.13861355, 1010.82919538,
         10.12911986,   32.85210359,   10.60108877])

Average of all of these columns

In [100]:
print("Mean of each column:", mean)
print("Average Elevation: ", mean[0].round(2))
print("Average Minimum Temperature: ", mean[1].round(2))
print("Average Maximum Temperature: ", mean[2].round(2))
print("Average Avg Temperature: ", mean[3].round(2))
print("Average Precipitation: ", mean[4].round(2))
print("Average Windspeed: ", mean[5].round(2))
print("Average Humidity: ", mean[6].round(2))
print("Average Pressure: ", mean[7].round(2))
print("Average Dew Point: ", mean[8].round(2))
print("Average Cloud Cover: ", mean[9].round(2))
print("Average Temperature Range: ", mean[10].round(2))

Mean of each column: [ 651.6568174    15.18648793   25.7875767    20.35544857    1.71780106
   12.90230655   58.13861355 1010.82919538   10.12911986   32.85210359
   10.60108877]
Average Elevation:  651.66
Average Minimum Temperature:  15.19
Average Maximum Temperature:  25.79
Average Avg Temperature:  20.36
Average Precipitation:  1.72
Average Windspeed:  12.9
Average Humidity:  58.14
Average Pressure:  1010.83
Average Dew Point:  10.13
Average Cloud Cover:  32.85
Average Temperature Range:  10.6


The Maximum Elevation Accordingly with the city, region and Date

In [101]:
index_of_max_elevation = np.argmax(numeric_data[:, 0])

max_total_elevation = numeric_data[index_of_max_elevation, 0]

season_of_max_elevation = full_data[index_of_max_elevation, 6]

city_with_max_elevation = full_data[index_of_max_elevation, 7]

region_with_max_elevation = full_data[index_of_max_elevation, 8]

date_of_max_elevation = full_data[index_of_max_elevation, 0]

print(f"The maximum elevation is: {full_data[index_of_max_elevation, 11]}")
print(f"The season of the maximum elevation is: {season_of_max_elevation}")
print(f"The city with the maximum elevation is: {city_with_max_elevation}")
print(f"The region with the maximum elevation is: {region_with_max_elevation}")
print(f"The date of maximum elevation is: {date_of_max_elevation}")

The maximum elevation is: 1680.0
The season of the maximum elevation is: Winter
The city with the maximum elevation is: Quetta
The region with the maximum elevation is: Balochistan
The date of maximum elevation is: 1/1/2014


The Minimum Elevation Accordingly with the city, region and Date

In [102]:
index_of_min_elevation = np.argmin(numeric_data[:, 0])

min_total_elevation = numeric_data[index_of_min_elevation, 0]

season_of_min_elevation = full_data[index_of_min_elevation, 6]

city_with_min_elevation = full_data[index_of_min_elevation, 7]

region_with_min_elevation = full_data[index_of_min_elevation, 8]

date_of_min_elevation = full_data[index_of_min_elevation, 0]

print(f"The minimum elevation is: {full_data[index_of_min_elevation, 11]}")
print(f"The season of the minimum elevation is: {season_of_min_elevation}")
print(f"The city with the minimum elevation is: {city_with_min_elevation}")
print(f"The region with the minimum elevation is: {region_with_min_elevation}")
print(f"The date of minimum elevation is: {date_of_min_elevation}")

The minimum elevation is: 8.0
The season of the minimum elevation is: Winter
The city with the minimum elevation is: Karachi
The region with the minimum elevation is: Sindh
The date of minimum elevation is: 1/1/2000


The Maximum Avg Temperature Accordingly with the city, region and Date

In [103]:
index_of_max_avg_temp = np.argmax(numeric_data[:, 3])

max_total_avg_temp = numeric_data[index_of_max_avg_temp, 3]

season_of_max_avg_temp = full_data[index_of_max_avg_temp, 6]

city_with_max_avg_temp = full_data[index_of_max_avg_temp, 7]

region_with_max_avg_temp = full_data[index_of_max_avg_temp, 8]

date_of_max_avg_temp = full_data[index_of_max_avg_temp, 0]

print(f"The maximum average temperature is: {full_data[index_of_max_avg_temp, 14]}")
print(f"The season of the maximum average temperature is: {season_of_max_avg_temp}")
print(f"The city with the maximum average temperature is: {city_with_max_avg_temp}")
print(f"The region with the maximum average temperature is: {region_with_max_avg_temp}")
print(f"The date of maximum average temperature is: {date_of_max_avg_temp}")

The maximum average temperature is: 39.2
The season of the maximum average temperature is: Summer
The city with the maximum average temperature is: Peshawar
The region with the maximum average temperature is: KPK
The date of maximum average temperature is: 6/26/2005


The Minimum Avg Temperature Accordingly with the city, region and Date

In [104]:
index_of_min_avg_temp = np.argmin(numeric_data[:, 3])

min_total_avg_temp = numeric_data[index_of_min_avg_temp, 3]

season_of_min_avg_temp = full_data[index_of_min_avg_temp, 6]

city_with_min_avg_temp = full_data[index_of_min_avg_temp, 7]

region_with_min_avg_temp = full_data[index_of_min_avg_temp, 8]

date_of_min_avg_temp = full_data[index_of_min_avg_temp, 0]

print(f"The minimum min temperature is: {full_data[index_of_min_avg_temp, 14]}")
print(f"The season of the minimum min temperature is: {season_of_min_avg_temp}")
print(f"The city with the minimum min temperature is: {city_with_min_avg_temp}")
print(f"The region with the minimum min temperature is: {region_with_min_avg_temp}")
print(f"The date of minimum min temperature is: {date_of_min_avg_temp}")

The minimum min temperature is: -15.6
The season of the minimum min temperature is: Winter
The city with the minimum min temperature is: Gilgit
The region with the minimum min temperature is: Gilgit-Baltistan
The date of minimum min temperature is: 1/19/2002


The Maximum Precipitation Accordingly with the city, region and Date

In [105]:
index_of_max_precipitation = np.argmax(numeric_data[:, 4])

max_total_precipitation = numeric_data[index_of_max_precipitation, 4]

season_of_max_precipitation = full_data[index_of_max_precipitation, 6]

city_with_max_precipitation = full_data[index_of_max_precipitation, 7]

region_with_max_precipitation = full_data[index_of_max_precipitation, 8]

date_of_max_precipitation = full_data[index_of_max_precipitation, 0]

print(f"The maximum precipitation is: {full_data[index_of_max_precipitation, 15]}")
print(f"The season of the maximum precipitation is: {season_of_max_precipitation}")
print(f"The city with the maximum precipitation is: {city_with_max_precipitation}")
print(f"The region with the maximum precipitation is: {region_with_max_precipitation}")
print(f"The date of maximum precipitation is: {date_of_max_precipitation}")

The maximum precipitation is: 201.1
The season of the maximum precipitation is: Summer
The city with the maximum precipitation is: Islamabad
The region with the maximum precipitation is: Punjab
The date of maximum precipitation is: 8/13/2007


The Minimum Precipitation Accordingly with the city, region and Date



In [106]:
index_of_min_precipitation = np.argmin(numeric_data[:, 4])

min_total_precipitation = numeric_data[index_of_min_precipitation, 4]

season_of_min_precipitation = full_data[index_of_min_precipitation, 6]

city_with_min_precipitation = full_data[index_of_min_precipitation, 7]

region_with_min_precipitation = full_data[index_of_min_precipitation, 8]

date_of_min_precipitation = full_data[index_of_min_precipitation, 0]

print(f"The minimum precipitation is: {full_data[index_of_min_precipitation, 15]}")
print(f"The season of the minimum precipitation is: {season_of_min_precipitation}")
print(f"The city with the minimum precipitation is: {city_with_min_precipitation}")
print(f"The region with the minimum precipitation is: {region_with_min_precipitation}")
print(f"The date of minimum precipitation is: {date_of_min_precipitation}")

The minimum precipitation is: 0.0
The season of the minimum precipitation is: Winter
The city with the minimum precipitation is: Islamabad
The region with the minimum precipitation is: Punjab
The date of minimum precipitation is: 1/1/2000


The Maximum Wind Speed Accordingly with the city, region and Date

In [107]:
index_of_max_windspd = np.argmax(numeric_data[:, 5])

max_total_windspd = numeric_data[index_of_max_windspd, 5]

season_of_max_windspd = full_data[index_of_max_windspd, 6]

city_with_max_windspd = full_data[index_of_max_windspd, 7]

region_with_max_windspd = full_data[index_of_max_windspd, 8]

date_of_max_windspd = full_data[index_of_max_windspd, 0]

print(f"The maximum Wind Speed is: {full_data[index_of_max_windspd, 16]}")
print(f"The season of the maximum Wind Speed is: {season_of_max_windspd}")
print(f"The city with the maximum Wind Speed is: {city_with_max_windspd}")
print(f"The region with the maximum Wind Speed is: {region_with_max_windspd}")
print(f"The date of maximum Wind Speed is: {date_of_max_windspd}")

The maximum Wind Speed is: 50.6
The season of the maximum Wind Speed is: Summer
The city with the maximum Wind Speed is: Karachi
The region with the maximum Wind Speed is: Sindh
The date of maximum Wind Speed is: 6/14/2000


The Minimum Wind Speed Accordingly with the city, region and Date

In [108]:
index_of_min_windspd = np.argmin(numeric_data[:, 5])

min_total_windspd = numeric_data[index_of_min_windspd, 5]

season_of_min_windspd = full_data[index_of_min_windspd, 6]

city_with_min_windspd = full_data[index_of_min_windspd, 7]

region_with_min_windspd = full_data[index_of_min_windspd, 8]

date_of_min_windspd = full_data[index_of_min_windspd, 0]

print(f"The minimum Wind Speed is: {full_data[index_of_min_windspd, 16]}")
print(f"The season of the minimum Wind Speed is: {season_of_min_windspd}")
print(f"The city with the minimum Wind Speed is: {city_with_min_windspd}")
print(f"The region with the minimum Wind Speed is: {region_with_min_windspd}")
print(f"The date of minimum Wind Speed is: {date_of_min_windspd}")

The minimum Wind Speed is: 0.8
The season of the minimum Wind Speed is: Spring
The city with the minimum Wind Speed is: Gilgit
The region with the minimum Wind Speed is: Gilgit-Baltistan
The date of minimum Wind Speed is: 4/10/2008


The Maximum Humidity Accordingly with the city, region and Date

In [109]:
index_of_max_humidity = np.argmax(numeric_data[:, 6])

max_total_humidity = numeric_data[index_of_max_windspd, 6]

season_of_max_humidity = full_data[index_of_max_windspd, 6]

city_with_max_humidity = full_data[index_of_max_windspd, 7]

region_with_max_humidity = full_data[index_of_max_windspd, 8]

date_of_max_humidity = full_data[index_of_max_windspd, 0]

print(f"The maximum humidity is: {full_data[index_of_max_humidity, 17]}")
print(f"The season of the maximum humidity is: {season_of_max_humidity}")
print(f"The city with the maximum humidity is: {city_with_max_humidity}")
print(f"The region with the maximum humidity is: {region_with_max_humidity}")
print(f"The date of maximum humidity is: {date_of_max_humidity}")

The maximum humidity is: 96.0
The season of the maximum humidity is: Summer
The city with the maximum humidity is: Karachi
The region with the maximum humidity is: Sindh
The date of maximum humidity is: 6/14/2000


The Minimum Humidity Accordingly with the city, region and Date

In [110]:
index_of_min_humidity = np.argmin(numeric_data[:, 6])

min_total_humidity = numeric_data[index_of_min_windspd, 6]

season_of_min_humidity = full_data[index_of_min_windspd, 6]

city_with_min_humidity = full_data[index_of_min_windspd, 7]

region_with_min_humidity = full_data[index_of_min_windspd, 8]

date_of_min_humidity = full_data[index_of_min_windspd, 0]

print(f"The minimum humidity is: {full_data[index_of_min_humidity, 17]}")
print(f"The season of the minimum humidity is: {season_of_min_humidity}")
print(f"The city with the minimum humidity is: {city_with_min_humidity}")
print(f"The region with the minimum humidity is: {region_with_min_humidity}")
print(f"The date of minimum humidity is: {date_of_min_humidity}")

The minimum humidity is: 6.0
The season of the minimum humidity is: Spring
The city with the minimum humidity is: Gilgit
The region with the minimum humidity is: Gilgit-Baltistan
The date of minimum humidity is: 4/10/2008


The Maximum Pressure Accordingly with the city, region and Date

In [111]:
index_of_max_pressure = np.argmax(numeric_data[:, 7])

max_total_pressure = numeric_data[index_of_max_pressure, 7]

season_of_max_pressure = full_data[index_of_max_pressure, 6]

city_with_max_pressure = full_data[index_of_max_pressure, 7]

region_with_max_pressure = full_data[index_of_max_pressure, 8]

date_of_max_pressure = full_data[index_of_max_pressure, 0]

print(f"The maximum pressure is: {full_data[index_of_max_pressure, 18]}")
print(f"The season of the maximum pressure is: {season_of_max_pressure}")
print(f"The city with the maximum pressure is: {city_with_max_pressure}")
print(f"The region with the maximum pressure is: {region_with_max_pressure}")
print(f"The date of maximum pressure is: {date_of_max_pressure}")

The maximum pressure is: 1042.9
The season of the maximum pressure is: Autumn
The city with the maximum pressure is: Gilgit
The region with the maximum pressure is: Gilgit-Baltistan
The date of maximum pressure is: 11/25/2003


The Minimum Pressure Accordingly with the city, region and Date

In [112]:
index_of_min_pressure = np.argmin(numeric_data[:, 7])

min_total_pressure = numeric_data[index_of_min_pressure, 7]

season_of_min_pressure = full_data[index_of_min_pressure, 6]

city_with_min_pressure = full_data[index_of_min_pressure, 7]

region_with_min_pressure = full_data[index_of_min_pressure, 8]

date_of_min_pressure = full_data[index_of_min_pressure, 0]

print(f"The minimum pressure is: {full_data[index_of_min_pressure, 18]}")
print(f"The season of the minimum pressure is: {season_of_min_pressure}")
print(f"The city with the minimum pressure is: {city_with_min_pressure}")
print(f"The region with the minimum pressure is: {region_with_min_pressure}")
print(f"The date of minimum pressure is: {date_of_min_pressure}")

The minimum pressure is: 989.7
The season of the minimum pressure is: Summer
The city with the minimum pressure is: Lahore
The region with the minimum pressure is: Punjab
The date of minimum pressure is: 6/13/2018


The Maximum Dew Point Accordingly with the city, region and Date

In [113]:
index_of_max_dew_point = np.argmax(numeric_data[:, 8])

max_total_dew_point = numeric_data[index_of_max_dew_point, 8]

season_of_max_dew_point = full_data[index_of_max_dew_point, 6]

city_with_max_dew_point = full_data[index_of_max_dew_point, 7]

region_with_max_dew_point = full_data[index_of_max_dew_point, 8]

date_of_max_dew_point = full_data[index_of_max_dew_point, 0]

print(f"The maximum dew point is: {full_data[index_of_max_dew_point, 19]}")
print(f"The season of the maximum dew point is: {season_of_max_dew_point}")
print(f"The city with the maximum dew point is: {city_with_max_dew_point}")
print(f"The region with the maximum dew point is: {region_with_max_dew_point}")
print(f"The date of maximum dew point is: {date_of_max_dew_point}")

The maximum dew point is: 27.5
The season of the maximum dew point is: Summer
The city with the maximum dew point is: Lahore
The region with the maximum dew point is: Punjab
The date of maximum dew point is: 7/31/2024


The Minimum Dew Point Accordingly with the city, region and Date

In [114]:
index_of_min_dew_point = np.argmin(numeric_data[:, 8])

min_total_dew_point = numeric_data[index_of_min_dew_point, 8]

season_of_min_dew_point = full_data[index_of_min_dew_point, 6]

city_with_min_dew_point = full_data[index_of_min_dew_point, 7]

region_with_min_dew_point = full_data[index_of_min_dew_point, 8]

date_of_min_dew_point = full_data[index_of_min_dew_point, 0]

print(f"The minimum dew point is: {full_data[index_of_min_dew_point, 19]}")
print(f"The season of the minimum dew point is: {season_of_min_dew_point}")
print(f"The city with the minimum dew point is: {city_with_min_dew_point}")
print(f"The region with the minimum dew point is: {region_with_min_dew_point}")
print(f"The date of minimum dew point is: {date_of_min_dew_point}")

The minimum dew point is: -28.7
The season of the minimum dew point is: Winter
The city with the minimum dew point is: Gilgit
The region with the minimum dew point is: Gilgit-Baltistan
The date of minimum dew point is: 1/2/2003


The Maximum Cloud Cover Accordingly with the city, region and Date

In [115]:
index_of_max_cloud_cover = np.argmax(numeric_data[:, 9])

max_total_cloud_cover = numeric_data[index_of_max_cloud_cover, 9]

season_of_max_cloud_cover = full_data[index_of_max_cloud_cover, 6]

city_with_max_cloud_cover = full_data[index_of_max_cloud_cover, 7]

region_with_max_cloud_cover = full_data[index_of_max_cloud_cover, 8]

date_of_max_cloud_cover = full_data[index_of_max_cloud_cover, 0]

print(f"The maximum cloud cover is: {full_data[index_of_max_cloud_cover, 20]}")
print(f"The season of the maximum cloud cover is: {season_of_max_cloud_cover}")
print(f"The city with the maximum cloud cover is: {city_with_max_cloud_cover}")
print(f"The region with the maximum cloud cover is: {region_with_max_cloud_cover}")
print(f"The date of maximum cloud cover is: {date_of_max_cloud_cover}")

The maximum cloud cover is: 100.0
The season of the maximum cloud cover is: Winter
The city with the maximum cloud cover is: Islamabad
The region with the maximum cloud cover is: Punjab
The date of maximum cloud cover is: 12/16/2000


The Minimum Cloud Cover Accordingly with the city, region and Date

In [116]:
index_of_min_cloud_cover = np.argmin(numeric_data[:, 9])

min_total_cloud_cover = numeric_data[index_of_min_cloud_cover, 9]

season_of_min_cloud_cover = full_data[index_of_min_cloud_cover, 6]

city_with_min_cloud_cover = full_data[index_of_min_cloud_cover, 7]

region_with_min_cloud_cover = full_data[index_of_min_cloud_cover, 8]

date_of_min_cloud_cover = full_data[index_of_min_cloud_cover, 0]

print(f"The minimum cloud cover is: {full_data[index_of_min_cloud_cover, 20]}")
print(f"The season of the minimum cloud cover is: {season_of_min_cloud_cover}")
print(f"The city with the minimum cloud cover is: {city_with_min_cloud_cover}")
print(f"The region with the minimum cloud cover is: {region_with_min_cloud_cover}")
print(f"The date of minimum cloud cover is: {date_of_min_cloud_cover}")

The minimum cloud cover is: 0.0
The season of the minimum cloud cover is: Winter
The city with the minimum cloud cover is: Islamabad
The region with the minimum cloud cover is: Punjab
The date of minimum cloud cover is: 1/1/2000


The Maximum Temperature Range Accordingly with the city, region and Date

In [117]:
index_of_max_temp_range = np.argmax(numeric_data[:, 10])

max_total_temp_range = numeric_data[index_of_max_temp_range, 9]

season_of_max_temp_range = full_data[index_of_max_temp_range, 6]

city_with_max_temp_range = full_data[index_of_max_temp_range, 7]

region_with_max_temp_range = full_data[index_of_max_temp_range, 8]

date_of_max_temp_range = full_data[index_of_max_temp_range, 0]

print(f"The maximum temperature range is: {full_data[index_of_max_temp_range, 21]}")
print(f"The season of the maximum temperature range is: {season_of_max_temp_range}")
print(f"The city with the maximum temperature range is: {city_with_max_temp_range}")
print(f"The region with the maximum temperature range is: {region_with_max_temp_range}")
print(f"The date of maximum temperature range is: {date_of_max_temp_range}")

The maximum temperature range is: 22.500000000000004
The season of the maximum temperature range is: Summer
The city with the maximum temperature range is: Peshawar
The region with the maximum temperature range is: KPK
The date of maximum temperature range is: 6/7/2002


The Minimum Temperature Range Accordingly with the city, region and Date

In [118]:
index_of_min_temp_range = np.argmin(numeric_data[:, 10])

min_total_temp_range = numeric_data[index_of_min_temp_range, 9]

season_of_min_temp_range = full_data[index_of_min_temp_range, 6]

city_with_min_temp_range = full_data[index_of_min_temp_range, 7]

region_with_min_temp_range = full_data[index_of_min_temp_range, 8]

date_of_min_temp_range = full_data[index_of_min_temp_range, 0]

print(f"The minimum temperature range is: {full_data[index_of_min_temp_range, 21]}")
print(f"The season of the minimum temperature range is: {season_of_min_temp_range}")
print(f"The city with the minimum temperature range is: {city_with_min_temp_range}")
print(f"The region with the minimum temperature range is: {region_with_min_temp_range}")
print(f"The date of minimum temperature range is: {date_of_min_temp_range}")

The minimum temperature range is: 0.7000000000000028
The season of the minimum temperature range is: Summer
The city with the minimum temperature range is: Karachi
The region with the minimum temperature range is: Sindh
The date of minimum temperature range is: 7/27/2015


# **Here are the key insights from the weather data analysis:**

## Overall Averages:

* *Average Elevation:* **651.66**
* *Average Minimum Temperature:* **15.19**
* *Average Maximum Temperature:* **25.79**
* *Average Average Temperature:* **20.36**
* *Average Precipitation:* **1.72**
* *Average Windspeed:* **12.9**
* *Average Humidity:* **58.14**
* *Average Pressure:* **1010.83**
* *Average Dew Point:* **10.13**
* *Average Cloud Cover:* **32.85**
* *Average Temperature Range:* **10.6**

## Extreme Values and Their Contexts:

* *Maximum Elevation:* **1680.0** in **Quetta**, **Balochistan**, on **1/1/2014 (Winter)**.
* *Minimum Elevation:* **8.0** in **Karachi**, **Sindh**, on **1/1/2000 (Winter)**.
* *Maximum Average Temperature:* **39.2** in **Peshawar**, **KPK**, on **6/26/2005 (Summer)**.
* *Minimum Average Temperature:* **-15.6** in **Gilgit**, **Gilgit-Baltistan**, on **1/19/2002 (Winter)**.
* *Maximum Precipitation:* **201.1** in **Islamabad**, **Punjab**, on **8/13/2007 (Summer)**.
* *Minimum Precipitation:* **0.0** in **Islamabad**, **Punjab**, on **1/1/2000 (Winter)**.
* *Maximum Wind Speed:* **50.6** in **Karachi**, **Sindh**, on **6/14/2000 (Summer)**.
* *Minimum Wind Speed:* **0.8** in **Gilgit**, **Gilgit-Baltistan**, on **4/10/2008 (Spring)**.
* *Maximum Humidity:* **96.0** in **Karachi**, **Sindh**, on **6/14/2000 (Summer)**.
* *Minimum Humidity:* **6.0** in **Gilgit**, **Gilgit-Baltistan**, on **4/10/2008 (Spring)**.
* *Maximum Pressure:* **1042.9** in **Gilgit**, **Gilgit-Baltistan**, on **11/25/2003 (Autumn)**.
* *Minimum Pressure:* **989.7** in **Lahore**, **Punjab**, on **6/13/2018 (Summer)**.
* *Maximum Dew Point:* **27.5** in **Lahore**, **Punjab**, on **7/31/2024 (Summer)**.
* *Minimum Dew Point:* **-28.7** in **Gilgit**, **Gilgit-Baltistan**, on **1/2/2003 (Winter)**.
* *Maximum Cloud Cover:* **100.0** in **Islamabad**, **Punjab**, on **12/16/2000 (Winter)**.
* *Minimum Cloud Cover:* **0.0** in **Islamabad**, **Punjab**, on **1/1/2000 (Winter)**.
* *Maximum Temperature Range:* **22.5** in **Peshawar**, **KPK**, on **6/7/2002 (Summer)**.
* *Minimum Temperature Range:* **0.7** in **Karachi**, **Sindh**, on **7/27/2015 (Summer)**.